### PR TITLE
[Api] upgrade taxon filter on product

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 1]
```diff
@@ -178,6 +182,14 @@ public function itsCurrentVariantShouldBeNamed(string $variantName): void
         );
     }
 
+    /**
+     * @Then I should see empty list of products
+     */
+    public function iShouldSeeEmptyListOfProducts(): void
+    {
+        Assert::same(0, $this->responseChecker->countTotalCollectionItems($this->client->getLastResponse()));
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:190]: `Assert::same($value, $expect, $message = '')` 😄 
```suggestion
        Assert::same($this->responseChecker->countTotalCollectionItems($this->client->getLastResponse()), 0);
```

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * It should be replaced with generic collection filter after fixed: https://github.com/api-platform/api-platform/issues/1868
+ */
+final class TaxonFilter extends AbstractContextAwareFilter
+{
+    /** @var IriConverter */
+    private $iriConverter;
+
+    /** @var ManagerRegistry  */
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:30]:
```suggestion
    /** @var ManagerRegistry */
```

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    protected $managerRegistry;
+
+    public function __construct(IriConverter $iriConverter, ManagerRegistry $managerRegistry)
+    {
+        parent::__construct($managerRegistry);
+        $this->iriConverter = $iriConverter;
+    }
+
+    protected function filterProperty(string $property, $value, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:39]: has `$value` any type? or is it mixed?

Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:39]: value is mixed, it is the value from the path, in this case, it is IRI (string) but it also can be int :)

File: features/product/viewing_products/viewing_products_from_specific_taxon.feature [hunk 4]
```diff
@@ -5,21 +5,20 @@ Feature: Viewing products from a specific taxon
     I want to be able to view products from a specific taxon
 
     Background:
-        Given the store has currency "Euro"
```
Comment (SirDomin) [inline features/product/viewing_products/viewing_products_from_specific_taxon.feature:8]: why is this removed?

Comment (AdamKasp) [reply features/product/viewing_products/viewing_products_from_specific_taxon.feature:8]: this step is unnecessary in these scenarios. so I deleted it :)

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 5]
```diff
@@ -61,7 +65,7 @@ public function iOpenProductPage(ProductInterface $product): void
     public function iBrowseProductsFromTaxon(TaxonInterface $taxon): void
     {
         $this->client->index();
-        $this->client->addFilter('productTaxons.taxon.code', $taxon->getCode());
+        $this->client->addFilter('productTaxons', $this->iriConverter->getIriFromItem($taxon));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:68]:
```suggestion
        $this->client->addFilter('taxon', $this->iriConverter->getIriFromItem($taxon));
```
?

File: src/Sylius/Behat/Service/Context/ChannelContext.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Behat\Service\Context;
+
+final class ChannelContext
+{
+
```
Comment (GSadee) [inline src/Sylius/Behat/Service/Context/ChannelContext.php:18]: Why do you need this empty context?

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 7]
```diff
@@ -40,6 +40,12 @@
             <tag name="api_platform.filter" />
         </service>
 
+        <service id="sylius.api.product_taxon_filter" class="Sylius\Bundle\ApiBundle\Filter\TaxonFilter">
+            <argument type="service" id="api_platform.iri_converter" />
+            <argument type="service" id="Doctrine\Common\Persistence\ManagerRegistry" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml:45]: To resolve the problem with psalm:
```suggestion
            <argument type="service" id="doctrine" />
```

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * It should be replaced with generic collection filter after fixed: https://github.com/api-platform/api-platform/issues/1868
+ */
+final class TaxonFilter extends AbstractContextAwareFilter
+{
+    /** @var IriConverterInterface */
+    private $iriConverter;
+
+    /** @var ManagerRegistry */
+    protected $managerRegistry;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:32]: Is it needed? As I see, you don't use this service.
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    public function __construct(IriConverterInterface $iriConverter, ManagerRegistry $managerRegistry)
+    {
+        parent::__construct($managerRegistry);
+        $this->iriConverter = $iriConverter;
+    }
+
+    public function filterProperty(string $property, $value, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
+    {
+        if ($property !== 'productTaxons') {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:42]: IMO the property should be simply `taxon`
```suggestion
        if ($property !== 'taxon') {
```

File: src/Sylius/Behat/Service/Context/ChannelContext.php (hunk 6 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Service/Context/ChannelContext.php:18]: I don't need it :)

**Review by arti0090 — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            'type' => 'string',
+            'required' => false,
+            'property' => null,
+            'swagger' => [
+                'name' =>'Product taxon filter',
+                'description' => 'Get a collection of product with chosen taxon'
+            ]
+        ];
+
+        return $description;
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:74]:
```suggestion
        return [
            'taxon' => [
                'type' => 'string',
                'required' => false,
                'property' => null,
                'swagger' => [
                    'name' =>'Product taxon filter',
                    'description' => 'Get a collection of product with chosen taxon',
                ],
            ],
        ];
```
```

File: src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+use ApiPlatform\Core\Api\IriConverterInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\AbstractContextAwareFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Doctrine\Persistence\ManagerRegistry;
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+
+/**
+ * It should be replaced with generic collection filter after fixed: https://github.com/api-platform/api-platform/issues/1868
+ */
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Filter/TaxonFilter.php:25]: now in this filter is more logic (we are getting child taxon) so this comment isn't true :)
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/spec/Filter/TaxonFilterSpec.php [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+use Symfony\Bridge\Doctrine\ManagerRegistry;
+
+final class TaxonFilterSpec extends ObjectBehavior
+{
+    function let(ManagerRegistry $managerRegistry, IriConverterInterface $iriConverter): void
+    {
+        $this->beConstructedWith($managerRegistry, $iriConverter);
+    }
+
+    function it_adds_taxon_filter_if_property_is_product_taxon(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Filter/TaxonFilterSpec.php:30]:
```suggestion
    function it_adds_taxon_filter_if_property_is_taxon(
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Adam! :1st_place_medal:
